File: gsea_exp.R
# gsea_exp.R
# Obtains gene expression binary scores based on pre-processed gene-set modules
# NOTE: This can be for either wilcoxon or ttest
# NOTE: No p.adjust is done!

library(data.table)
library(reshape2)
library(parallel)

# Functions
Function_exp_combat <- function(exp.matrix.1, exp.matrix.2) {
  #Function to normalize batch effects across two expression matrices
  #INPUT: 2 expression matrices with columns as samples and rows as genes
  #OUPOUT: 2 expression matrices, HOWEVER, genes that are not common across expression datasets will be removed
  
  require(sva)
  require(pamr)
  require(limma)
  
  # Remove null standard gene deviations from each matrix first
  exp.matrix.1 <- exp.matrix.1[apply(exp.matrix.1, 1, sd)!=0, ]
  exp.matrix.2 <- exp.matrix.2[apply(exp.matrix.2, 1, sd)!=0, ]

  common.genes <- intersect( unique(rownames(exp.matrix.1)) , 
                             unique(rownames(exp.matrix.2)))
  print(length(common.genes))

  col.1 <- colnames(exp.matrix.1)
  col.2 <- colnames(exp.matrix.2)
  
  m.1 <- exp.matrix.1[common.genes,]
  m.2 <- exp.matrix.2[common.genes,]
  m.both <- cbind(m.1, m.2)
  
  # Filter out zero sd
  m.both <- m.both[apply(m.both, 1, sd)!=0, ]
  
  # Continue
  
  batch <- c( rep("m1", length(col.1)) , 
              rep("m2", length(col.2)) )
  
  modcombat <- model.matrix(~1, data.frame(1:length(batch)) )
  
  combat.m <- ComBat(dat=m.both, batch=batch, mod=modcombat, par.prior = T, prior.plots = F)
  
  exp.matrix.1 <- combat.m[,col.1]
  exp.matrix.2 <- combat.m[,col.2]
  
  return(list(EXP.1=exp.matrix.1 , EXP.2=exp.matrix.2))
}

Function_process_ttest <- function(){
	nodes<-detectCores()
	print(nodes)
	cl<-makeCluster(nodes)
	setDefaultCluster(cl)
	clusterExport(cl, varlist=c("data.table", "as.data.table", "exp_samples", "gene_sets", "in_exp", "gsea"),envir=environment())

	main_table <- parApply(cl, gene_sets, 1, function(g) {

		name    <- paste0(g[1], "$", g[2])
		genes_1 <- gsea[Gene_set==g[1],]$genes
		genes_2 <- gsea[Gene_set==g[2],]$genes

		pvals <- sapply(exp_samples, function(s){

			genes_1_exp <- as.vector(in_exp[genes_1, s])
			genes_2_exp <- as.vector(in_exp[genes_2, s])

			return(t.test(genes_1_exp, genes_2_exp, alternative="greater")$p.value)
		})

		return(data.table(gs = name, sample = exp_samples, pvals))
	})

	stopCluster(cl)

	main_table <- do.call(rbind, main_table)
	return(main_table)
}

Function_process_wilcoxontest <- function(){
	nodes<-detectCores()
	print(nodes)
	cl<-makeCluster(nodes)
	setDefaultCluster(cl)
	clusterExport(cl, varlist=c("data.table", "as.data.table", "exp_samples", "gene_sets", "in_exp", "gsea"),envir=environment())

	main_table <- parApply(cl, gene_sets, 1, function(g) {

		name    <- paste0(g[1], "$", g[2])
		genes_1 <- gsea[Gene_set==g[1],]$genes
		genes_2 <- gsea[Gene_set==g[2],]$genes

		pvals <- sapply(exp_samples, function(s){

			genes_1_exp <- as.vector(in_exp[genes_1, s])
			genes_2_exp <- as.vector(in_exp[genes_2, s])

			return(wilcox.test(genes_1_exp, genes_2_exp, alternative="greater")$p.value)
		})

		return(data.table(gs = name, sample = exp_samples, pvals))
	})

	stopCluster(cl)

	main_table <- do.call(rbind, main_table)
	return(main_table)
}

# Load arguments
args      <- commandArgs(trailingOnly = TRUE)
exp_type  <- args[1]
base      <- args[2]
gsea_type <- args[3]
both      <- args[4]
test      <- args[5] # wilcoxon or ttest
bn        <- args[6] # against which other dataset is being normalize to (or None)

# Load files
in_folder <- "/tigress/zamalloa/"
if (exp_type=="cgp"){
	in_exp <- readRDS(paste0(in_folder,"CGP_FILES/083016_cgp_exp.rds"))
} else if (exp_type=="bortezomib_a"){
	in_exp <- readRDS(paste0(in_folder, "OBJECTS/030417_GEE_BORTEZOMIB_DEXAMETHASONE.rds"))[["exp_table_a"]]
} else if (exp_type=="bortezomib_b"){
	in_exp <- readRDS(paste0(in_folder, "OBJECTS/030417_GEE_BORTEZOMIB_DEXAMETHASONE.rds"))[["exp_table_b"]]
} else if (exp_type=="docetaxel"){
	in_exp <- readRDS(paste0(in_folder, "OBJECTS/030217_GEE_DOCETAXEL.rds"))[["exp_table"]]
} else if (exp_type=="erlotinib"){
	in_exp <- readRDS(paste0(in_folder, "OBJECTS/030417_GEE_ERLOTINIB.rds"))[["exp_table"]]
} else if (exp_type=="cisplatin"){
	in_exp <- readRDS(paste0(in_folder, "OBJECTS/030217_GEE_CISPLATIN.rds"))[["exp_table"]]
}

# Do we need to apply batch normalization
if (bn!="None"){
	if (bn=="cgp"){
		bn_exp <- readRDS(paste0(in_folder,"CGP_FILES/083016_cgp_exp.rds"))
	} else if (bn=="bortezomib_a"){
		bn_exp <- readRDS(paste0(in_folder, "OBJECTS/030417_GEE_BORTEZOMIB_DEXAMETHASONE.rds"))[["exp_table_a"]]
	} else if (bn=="bortezomib_b"){
		bn_exp <- readRDS(paste0(in_folder, "OBJECTS/030417_GEE_BORTEZOMIB_DEXAMETHASONE.rds"))[["exp_table_b"]]
	} else if (bn=="docetaxel"){
		bn_exp <- readRDS(paste0(in_folder, "OBJECTS/030217_GEE_DOCETAXEL.rds"))[["exp_table"]]
	} else if (bn=="erlotinib"){
		bn_exp <- readRDS(paste0(in_folder, "OBJECTS/030417_GEE_ERLOTINIB.rds"))[["exp_table"]]
	} else if (bn=="cisplatin"){
		bn_exp <- readRDS(paste0(in_folder, "OBJECTS/030217_GEE_CISPLATIN.rds"))[["exp_table"]]
	}

	print("Batch normalizing")
	in_exp <- Function_exp_combat(in_exp, bn_exp)[["EXP.1"]]
}


# Load gseas
gsea    <- lapply(strsplit(gsea_type, "_")[[1]], function(x) {
	return(fread(paste0("GSEA_FILES/",x, "_sets")))
})
gsea    <- do.call(rbind, gsea)

# Pre-process
exp_samples <- colnames(in_exp)
exp_genes   <- rownames(in_exp)
print(nrow(gsea))
gsea        <- gsea[genes %in% exp_genes,]
print(nrow(gsea))

# Do we filter by base
if (base=="None"){
	gsea[,N:=length(unique(genes)), by="Gene_set"]
	gene_sets <- unique(gsea[N >= 100,]$Gene_set)
	gene_sets <- t(combn(gene_sets, 2))

	if (both == "T"){
		print("both one-tailed used")
		gene_sets <- rbind(gene_sets, gene_sets[,c(2,1)])
	}
	
} else {
	base      <- fread(paste0(in_folder, "GSEA_FILES/", gsea_type, "_gsea_", base, "_both_", both, "_pvals"))
	gene_sets <- unique(base$gs)
	gene_sets <- t(sapply(gene_sets, function(x) strsplit(x, "$", fixed = T)[[1]]))
}

print(c(gsea_type, nrow(gene_sets)))

# Process depending on test
if (test=="ttest"){
	print("performing ttest")
	main_table <- Function_process_ttest()
} else if (test=="wilcoxon"){
	print("performing wilcoxon")
	main_table <- Function_process_wilcoxontest()
}

print("Done calculating")

# saveRDS(main_table, "/tigress/zamalloa/GSEA_FILES/all_gsea_cgp_pvals.rds")
#saveRDS(main_table, paste0("/tigress/zamalloa/GSEA_FILES/",g,"_gsea_cgp_pvals.rds"))
file_out <- paste0(in_folder, "GSEA_FILES/", gsea_type, "_gsea_", exp_type, "_", test,"_both_", both, "_bn_",bn,"_pvals")
write.table(main_table, file_out, quote = F, sep="\t", row.names = F, col.names = T)
print("Done writing")

File: gsea_exp.py
# gsea_exp.py
# cgp_site_gmv.R was used to build cgp_site_gmv_exp
# tcga expression sets (including tcga_site_gmv were constructed using tcga_exp.R)

import sys
import pandas as pd
import numpy as np
import itertools
from scipy.stats import mannwhitneyu
from scipy.stats import ttest_ind
from joblib import Parallel, delayed

def mann_pval(gs):

	name    = gs[0] + "$" + gs[1]
	genes_1 = gsea_dict[gs[0]]
	genes_2 = gsea_dict[gs[1]]

	genes_1_index = [i for i,n in enumerate(genes) if n in genes_1]
	genes_2_index = [i for i,n in enumerate(genes) if n in genes_2]

	gs_pvals = []
	for s in xrange(len(samples)):

		genes_1_exp = exp_matrix[genes_1_index, s]
		genes_2_exp = exp_matrix[genes_2_index, s]

		# Perform wilcoxon test
		pvals = mannwhitneyu(genes_1_exp, genes_2_exp, alternative="greater")[1]
		# pvals = ttest_ind(genes_1_exp, genes_2_exp, equal_var=False)[1]

		gs_pvals = gs_pvals + [pvals]

	return({"sample":samples, "gs":name, "pvals":gs_pvals})

# Load arguments
exp_type   = sys.argv[1]
base       = sys.argv[2] # Base can be None or cgp/ctrp
gsea_type  = sys.argv[3] # ("h", "c1", "c2.cp.biocarta", "c3", "cancer") - OR - joined combination (i.e. h_c1)
both       = sys.argv[4] # T/F Both tests, not 2 sided, but one sided on both sides
ext_gmv    = sys.argv[5] # T/F apply gmv normalization in this script
drug       = sys.argv[6] # Cisplatin, Bortezomib (unlike bortezomib_a in 'exp_type')

# Load files
in_mac     = "/Users/jzamalloa/Documents/Rotation/PIPELINES/"
in_tiger   = "/tigress/zamalloa/"
in_lab     = "/home/zamalloa/Documents/Rotation/PIPELINES/"
in_folder  = in_lab

gsea       = pd.DataFrame()
for g in gsea_type.split("_"):
	gsea       = pd.concat([gsea,
							pd.read_csv(in_folder + "GSEA_FILES/" + g + "_sets", sep="\t")]) 

if exp_type=="ctrp":
	in_exp    = pd.read_csv(in_folder + "CTRP_FILES/ctrp_exp_2.1", sep="\t")
else:
	in_exp    = pd.read_csv(in_folder + "CGP_FILES/" + exp_type + "_exp", sep="\t")	

	# Selecting relevant samples
	if "cgp" not in exp_type:
		target    = pd.read_csv(in_folder + "CGP_FILES/" + exp_type + "_target", sep="\t")
		target    = target[target.Compound==drug]
		in_exp    = in_exp.loc[:, ["rn"]+list(target["cell_name"])]

# Pre-process
exp_matrix = in_exp.iloc[:, 1:].as_matrix()
exp_mean   = np.mean(exp_matrix, axis=1).reshape(exp_matrix.shape[0], 1)
exp_var    = np.var(exp_matrix, axis=1, ddof=1).reshape(exp_matrix.shape[0], 1)

if ext_gmv == "T":

	print("Applying gmv normalization")
	exp_matrix = (exp_matrix - exp_mean) / exp_var

samples    = list(in_exp)[1:]
genes      = list(in_exp["rn"])

print(gsea.shape)
gsea       = gsea[gsea.genes.isin(genes)]
gsea_dict  = {i:list(j) for i,j in gsea.groupby("Gene_set")["genes"]}
print(gsea.shape)

# Do we filter by a base?
if base=="None":
	gene_sets  = gsea.groupby(["Gene_set"]).size().reset_index(name="count")
	print(gene_sets)
	gene_sets  = gene_sets[gene_sets["count"] >= 100] #Changed back to 100!!!
	print(gene_sets)
	# gene_sets  = gene_sets[gene_sets["count"] <= 150]
	# print(gene_sets)
	gene_sets  = [i for i in itertools.combinations(list(gene_sets["Gene_set"]),2)]
	
	if both=="T":
		print("both one-tailed used")
		gene_r    = [(i[1], i[0]) for i in gene_sets]
		gene_sets = gene_sets + gene_r
else:
	base       = pd.read_csv(in_folder + "GSEA_FILES/" + gsea_type + "_gsea_" + base + "_both_" + both + "_pvals", sep="\t")
	gene_sets  = list(set(base["gs"]))
	gene_sets  = [(i.split("$")[0], i.split("$")[1]) for i in gene_sets]

print(gsea_type, len(gene_sets))

main_dict = Parallel(n_jobs=40)(delayed(mann_pval)(i) for i in gene_sets)

print("Done calculating")
# Write to file
main_dict = pd.concat([pd.DataFrame(i) for i in main_dict])
file_out  = in_folder + "GSEA_FILES/" + gsea_type + "_gsea_"+ exp_type + "_both_" + both + "_ext_gmv_" + ext_gmv + "_pvals"

main_dict.to_csv(file_out, sep="\t", header=True, index=False)

print ("Done writing")